mungegithub/mungers/publisher.go

```diff
@@ -180,64 +220,128 @@ find %s -depth -maxdepth 1 \( \
+
 		if err = os.Chdir(dstDir); err != nil {
-			glog.Errorf("Failed to chdir to %s.\nError: %s", dstDir, err)
+			p.plog.Errorf("Failed to chdir to %s.\nError: %s", dstDir, err)
 			errlist = append(errlist, err)
+			if repoRules.abortAllRepos {
+				p.plog.Errorf("abort the current publishing cycle")
+				break Repos
+			}
 			continue Repos
 		}
-		// construct the repo's branches and subdirs
-		for src, dst := range rules.srcToDst {
-			var commitMessage = "published by bot\n(https://github.com/kubernetes/test-infra/tree/master/mungegithub)\n\n"
-			if err = exec.Command("git", "checkout", dst.branch).Run(); err != nil {
-				glog.Errorf("Failed to checkout branch %s.\nError: %s", dst.branch, err)
-				errlist = append(errlist, err)
-				continue Repos
-			}
-			dstRepoRoot := filepath.Join(p.baseDir, dst.repo)
-			snippet, err := construct(p.baseDir, p.githubConfig.Org, src, dst)
-			if err != nil {
-				glog.Errorf("Failed to construct %s.\nError: %s", dstRepoRoot, err)
-				errlist = append(errlist, err)
-				continue Repos
-			} else {
-				commitMessage += snippet
-				glog.Infof("Successfully construct %s", filepath.Join(dstRepoRoot, dst.dir))
+		// work on the branches of the repo
+		for src, dst := range repoRules.srcToDst {
+			// currently apimachinery uses its own script
+			if len(repoRules.publishScript) != 0 {
+				// TODO: pass in the branches
+				cmd := exec.Command(repoRules.publishScript, p.githubConfig.Token(), p.netrcDir)
+				output, err := cmd.CombinedOutput()
+				if err != nil {
+					p.plog.Errorf("Failed to publish %s.\nOutput: %s\nError: %s", dst, output, err)
+					errlist = append(errlist, err)
+					if repoRules.abortAllRepos {
+						p.plog.Errorf("abort the current publishing cycle")
+						break Repos
+					}
+					continue
+				}
+				p.plog.Infof("Successfully publish %s: %s", dst, output)
+				continue
 			}
 
-			// publish the destination branch
-			cmd := exec.Command("/publish.sh", filepath.Join(dstRepoRoot, dst.dir), dst.branch, p.githubConfig.Token(), p.netrcDir, strings.TrimSpace(commitMessage))
-			output, err := cmd.CombinedOutput()
+			// currently client-go takes this path.
+			// TODO: retire publishClientGo, centralize the code in client-go's publishScript.
+			err := p.publishClientGo(src, dst)
 			if err != nil {
-				glog.Errorf("Failed to publish %s.\nOutput: %s\nError: %s", dst, output, err)
 				errlist = append(errlist, err)
-				continue Repos
-			} else {
-				glog.Infof("Successfully publish %s: %s", dst, output)
+				if repoRules.abortAllRepos {
+					p.plog.Errorf("abort the current publishing cycle")
+					break Repos
+				}
+				continue
 			}
 		}
 	}
-	return errors.NewAggregate(errlist)
+
+	// create an issue or update an existing issue to report the failed
+	// publisher run
+	aggregated := errors.NewAggregate(errlist)
+	if len(errlist) != 0 {
+		log, err := readLog(p.logDir)
+		if err != nil {
+			glog.Errorf("failed to extract log: %s", err)
+			return aggregated
```

Comment:
Good point. Changed to let the robot file an issue stating "cannot extract log".